File: problems/strings/Binary_sum.js
/**
 * Given two binary strings A and B, return their sum (also a binary string).
 *
 * Example:
 * Input: A = "100", B = "11"
 * Output: "111"
 */

function soln(A, B) {
  let aArr = Array.from(A);
  let bArr = Array.from(B);
  let k = 0;
  while (k < aArr.length && aArr[k] === "0") {
    k++;
  }
  if (k > 0) {
    aArr.splice(0, k);
  }
  k = 0;
  while (k < bArr.length && bArr[k] === "0") {
    k++;
  }
  if (k > 0) {
    bArr.splice(0, k);
  }

  const n = aArr.length;
  const m = bArr.length;

  let i = n - 1;
  let j = m - 1;
  let carry = 0;
  let finalSum = "";
  while (i >= 0 || (j >= 0) | (carry > 0)) {
    let sum = carry;
    if (i >= 0) {
      sum += Number(aArr[i]);
      i--;
    }
    if (j >= 0) {
      sum += Number(bArr[j]);
      j--;
    }

    if (sum > 1) {
      carry = 1;
      sum = sum % 2;
    } else {
      carry = 0;
    }

    finalSum = sum.toString() + finalSum;
  }

  return finalSum;
}

console.log(soln("100", "11")); // "111"
console.log(soln("1101", "111")); // "10100"
console.log(soln("01001001", "0110101")); // "10111010"

File: src/main/java/com/dsa/problems/strings/Binary_sum.java
/**
 * Given two binary strings A and B, return their sum (also a binary string).
 *
 * Example:
 * Input: A = "100", B = "11"
 * Output: "111"
 */

package com.dsa.problems.strings;

import java.util.ArrayList;

public class Binary_sum {
  public static String soln(String A, String B) {
    ArrayList<Character> aArr = new ArrayList<>();
    ArrayList<Character> bArr = new ArrayList<>();

    for (int i = 0; i < A.length(); i++) {
      aArr.add(A.charAt(i));
    }
    for (int i = 0; i < B.length(); i++) {
      bArr.add(B.charAt(i));
    }

    int k = 0;
    while (k < aArr.size() && aArr.get(k) == '0') {
      k++;
    }
    if (k > 0) {
      aArr.subList(0, k).clear();
    }

    k = 0;
    while (k < bArr.size() && bArr.get(k) == '0') {
      k++;
    }
    if (k > 0) {
      bArr.subList(0, k).clear();
    }

    int n = aArr.size();
    int m = bArr.size();

    int i = n - 1;
    int j = m - 1;
    int carry = 0;
    StringBuilder finalSum = new StringBuilder();

    while (i >= 0 || j >= 0 || carry > 0) {
      int sum = carry;
      if (i >= 0) {
        sum += Character.getNumericValue(aArr.get(i));
        i--;
      }
      if (j >= 0) {
        sum += Character.getNumericValue(bArr.get(j));
        j--;
      }

      carry = sum / 2;
      sum = sum % 2;

      finalSum.insert(0, sum);
    }

    return finalSum.toString();
  }

  public static void main(String[] args) {
    System.out.println(soln("100", "11")); // "111"
    System.out.println(soln("1101", "111")); // "10100"
    System.out.println(soln("01001001", "0110101")); // "1011110"
  }
}
